presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MergeIdenticalWindows.java

```diff
@@ -93,33 +93,77 @@ protected PlanNode visitPlan(PlanNode node, RewriteContext<Multimap<Specificatio
             checkState(node.getPrePartitionedInputs().isEmpty() && node.getPreSortedOrderPrefix() == 0, "MergeIdenticalWindows should be run before AddExchanges");
+            checkState(node.getFrames().values().size() == 1, "More than one frame per WindowNode is not yet supported");
 
             return context.rewrite(
                     node.getSource(),
-                    ImmutableListMultimap.<Specification, WindowNode>builder()
-                            .put(node.getSpecification(), node) // Add the current window first so that it gets precedence in iteration order
+                    ImmutableListMultimap.<SpecificationAndFrame, WindowNode>builder()
+                            .put(new SpecificationAndFrame(node), node) // Add the current window first so that it gets precedence in iteration order
                             .putAll(context.get())
                             .build());
         }
 
-        private static WindowNode collapseWindows(PlanNode source, Specification specification, Collection<WindowNode> windows)
+        private static WindowNode collapseWindows(PlanNode source, SpecificationAndFrame specification, Collection<WindowNode> windows)
         {
             WindowNode canonical = windows.iterator().next();
+            WindowNode.Frame frame = canonical.getFrames().values().iterator().next(); // asserted in #visitWindow already
             return new WindowNode(
                     canonical.getId(),
                     source,
-                    specification,
+                    specification.specification,
                     windows.stream()
                             .map(WindowNode::getWindowFunctions)
                             .flatMap(map -> map.entrySet().stream())
                             .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue)),
+                    windows.stream()
+                            .map(WindowNode::getWindowFunctions)
+                            .flatMap(map -> map.entrySet().stream())
+                            .collect(Collectors.toMap(Map.Entry::getValue, ignore -> frame)),
                     canonical.getHashSymbol(),
                     canonical.getPrePartitionedInputs(),
                     canonical.getPreSortedOrderPrefix());
         }
+
+        static final class SpecificationAndFrame
+        {
+            private final WindowNode.Specification specification;
+            private final WindowNode.Frame frame;
+
+            SpecificationAndFrame(WindowNode.Specification specification, WindowNode.Frame frame)
+            {
+                this.specification = specification;
+                this.frame = frame;
+            }
+
+            SpecificationAndFrame(WindowNode node)
+            {
+                this(node.getSpecification(),
+                        node.getFrames().values().iterator().next()); // asserted in #visitWindow already
```

Comment:
move this into the callsite of the surrounding class since it encodes assumptions from that context (and get rid of this constructor)